### DISCUSSION TITLE
Accessing the np.array inside Py5Vector?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#272] [Q&A]
My friend @melissawm was helping me with the particle simulations and we wondered about how to broadcast operations could be done with the Py5Vector components.

Would it be possible to add ufuncs to Py5Vector operations?
https://numpy.org/doc/stable/reference/ufuncs.html

**Comment by hx2A** [comment]
You can access the np.array inside Py5Vector with the [data](https://py5coding.org/reference/py5vector_data.html) property (ie `v1.data`). Broadcast operations should work right now. Can you give an example of what you are trying to do that you can't do now? You can pass Py5Vector instances to numpy functions (ie `np.sin(v1)`) and get a numpy array back.

> Would it be possible to add ufuncs to Py5Vector operations?

I never totally understood what ufuncs were but just went through the documentation you linked. It seems `np.ufuncs` is a class and numpy functions like `np.sin` are instances of that class? It seems my understanding had been tripped up by the lower case `u`. Now it makes a bit more sense. 

Not sure what you are suggesting here. Are you suggesting that Py5Vector methods be instances of ufuncs? I'm open to making changes here, especially ones that promote tighter integration with numpy.

**Reply by villares** [reply]
Melissa can explain, she'll be watching this thread.  I'm just starting to get a grasp of it. It might be related to having a np.array of Py5Vectors and to be able to broadcast to the vector components... without having to move the data to a "pure" np.array structure for operations and then moving it back to Py5Vector objects.

**Reply by hx2A** [reply]
> Melissa can explain, she'll be watching this thread. I'm just starting to get a grasp of it. It might be related to having a np.array of Py5Vectors and to be able to broadcast to the vector components... without having to move the data to a "pure" np.array structure for operations and then moving it back to Py5Vector objects.

Sounds interesting! If this is possible, I'd be eager to learn how to do it!
